src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py

```diff
@@ -0,0 +1,248 @@
+
+"""Module to define how to subscribe and fetch component data."""
+
+from __future__ import annotations
+
+import logging
+import math
+from abc import ABC, abstractmethod
+from typing import Any, Generic, Iterable, Optional, Set, TypeVar
+
+from frequenz.channels import ChannelClosedError, Receiver
+
+from ..._internal.asyncio import AsyncConstructible
+from ...actor._data_sourcing.microgrid_api_source import (
+    _BatteryDataMethods,
+    _InverterDataMethods,
+)
+from ...microgrid import get as get_microgrid
+from ...microgrid.component import (
+    BatteryData,
+    ComponentCategory,
+    ComponentData,
+    ComponentMetricId,
+    InverterData,
+)
+from ._component_metrics import ComponentMetricsData
+
+_logger = logging.getLogger(__name__)
+
+T = TypeVar("T", bound=ComponentData)
+
+
+class ComponentMetricFetcher(AsyncConstructible, ABC):
+    """Define how to subscribe for and fetch the component metrics data."""
+
+    _component_id: int
+    _metrics: Iterable[ComponentMetricId]
+
+    @classmethod
+    async def async_new(
+        cls, component_id: int, metrics: Iterable[ComponentMetricId]
+    ) -> Self:  # pylint: disable=undefined-variable
+        """Create an instance of this class.
+
+        Subscribe for the given component metrics and return them if method
+        `fetch_next` is called.
+
+        Args:
+            component_id: component id
+            metrics: metrics that should be fetched from this component.
+
+        Returns:
+            This class instance.
+        """
+        self: ComponentMetricFetcher = ComponentMetricFetcher.__new__(cls)
+        self._component_id = component_id
+        self._metrics = metrics
+        return self
+
+    @abstractmethod
+    async def fetch_next(self) -> Optional[ComponentMetricsData]:
+        """Fetch metrics for this component."""
+
+
+class LatestMetricsFetcher(ComponentMetricFetcher, Generic[T], ABC):
+    """Subscribe for the latest component data and extract the needed metrics."""
+
+    _receiver: Receiver[T]
+
+    @classmethod
+    async def async_new(
+        cls,
+        component_id: int,
+        metrics: Iterable[ComponentMetricId],
+    ) -> Self:  # pylint: disable=undefined-variable:
+        """Create instance of this class.
+
+        Subscribe for the requested component data and fetch only the latest component
+        metrics.
+
+        Args:
+            component_id: component id
+            metrics: metrics
+
+        Raises:
+            ValueError: If any requested metric id is not supported.
+
+        Returns:
+            This class instance
+        """
+        self: LatestMetricsFetcher[T] = await super().async_new(component_id, metrics)
+
+        for metric in metrics:
+            # pylint: disable=protected-access
+            if metric not in self._supported_metrics():
+                category = self._component_category()
+                raise ValueError(f"Metric {metric} not supported fro the {category}")
```

Comment:
Small typo + minor improvement suggestion:

```suggestion
                raise ValueError(f"Metric {metric} not supported for {category}")
```

Reply:
Done